packages/dds/tree/src/feature-libraries/flex-map-tree/mapTreeNode.ts

```diff
@@ -0,0 +1,744 @@
+ */
+export class MapTreeFieldNode<TSchema extends FlexFieldNodeSchema>
+	extends MapTreeNode<TSchema>
+	implements FlexTreeFieldNode<TSchema>
+{
+	public get content(): FlexTreeUnboxField<TSchema["info"]> {
+		const field = this.tryGetField(EmptyKey);
+		if (field === undefined) {
+			return undefined as FlexTreeUnboxField<TSchema["info"]>;
+		}
+		return unboxedField(field, EmptyKey, this.mapTree, this);
+	}
+
+	public get boxedContent(): FlexTreeTypedField<TSchema["info"]> {
+		const field = this.mapTree.fields.get(EmptyKey) ?? [];
+		return getOrCreateField(
+			this,
+			EmptyKey,
+			field,
+			this.schema.info,
+		) as unknown as FlexTreeTypedField<TSchema["info"]>;
+	}
+
+	public override getBoxed(key: string): FlexTreeTypedField<TSchema["info"]> {
+		return super.getBoxed(key) as FlexTreeTypedField<TSchema["info"]>;
+	}
+}
+
+/**
+ * The implementation of a map node created by {@link getOrCreateMapTreeNode}.
+ */
+export class MapTreeMapNode<TSchema extends FlexMapNodeSchema>
+	extends MapTreeNode<TSchema>
+	implements FlexTreeMapNode<TSchema>
+{
+	public get size(): number {
+		return this.mapTree.fields.size;
+	}
+
+	public has(key: string): boolean {
+		return this.tryGetField(brand(key)) !== undefined;
+	}
+
+	public get(key: string): FlexTreeUnboxField<TSchema["info"]> {
+		const field = this.tryGetField(brand(key));
+		if (field === undefined) {
+			return undefined as FlexTreeUnboxField<TSchema["info"]>;
+		}
+		return unboxedField(field, brand(key), this.mapTree, this);
+	}
+
+	public keys(): IterableIterator<FieldKey> {
+		return this.mapTree.fields.keys();
+	}
+
+	public values(): IterableIterator<FlexTreeUnboxField<TSchema["info"], "notEmpty">> {
+		return mapIterable(
+			this.mapTree.fields.keys(),
+			(key) =>
+				unboxedField(
+					this.tryGetField(key) ?? fail("Unexpected empty map field"),
+					key,
+					this.mapTree,
+					this,
+				) as FlexTreeUnboxField<TSchema["info"], "notEmpty">,
+		);
+	}
+
+	public entries(): IterableIterator<
+		[FieldKey, FlexTreeUnboxField<TSchema["info"], "notEmpty">]
+	> {
+		return mapIterable(this.mapTree.fields.keys(), (key) => [
+			key,
+			unboxedField(
+				this.tryGetField(key) ?? fail("Unexpected empty map field"),
+				key,
+				this.mapTree,
+				this,
+			) as FlexTreeUnboxField<TSchema["info"], "notEmpty">,
+		]);
+	}
+
+	public forEach(
+		callbackFn: (
+			value: FlexTreeUnboxField<TSchema["info"], "notEmpty">,
+			key: FieldKey,
+			map: FlexTreeMapNode<TSchema>,
+		) => void,
+		thisArg?: unknown,
+	): void {
+		const fn = thisArg !== undefined ? callbackFn.bind(thisArg) : callbackFn;
+		for (const [key, value] of this.entries()) {
+			fn(value, key, this);
+		}
+	}
+
+	public override getBoxed(key: string): FlexTreeTypedField<TSchema["info"]> {
+		return super.getBoxed(key) as FlexTreeTypedField<TSchema["info"]>;
+	}
+
+	public set(key: string, value: FlexibleFieldContent<TSchema["info"]> | undefined): void {
+		throw unsupportedError("Setting a map entry");
```

Comment:
We should document why `set` and `delete` are unsupported here.

Reply:
Done.